src/Microsoft.ML.EntryPoints/JsonUtils/JsonManifestUtils.cs

```diff
@@ -153,15 +153,24 @@ private static JArray BuildInputManifest(IExceptionContext ectx, Type inputType,
 
             // Instantiate a value of the input, to pull defaults out of.
             var defaults = Activator.CreateInstance(inputType);
-
+            var collectedFields = new HashSet<string>();
             var inputs = new List<KeyValuePair<Double, JObject>>();
             foreach (var fieldInfo in inputType.GetFields(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance))
             {
                 var inputAttr = fieldInfo.GetCustomAttributes(typeof(ArgumentAttribute), false).FirstOrDefault() as ArgumentAttribute;
                 if (inputAttr == null || inputAttr.Visibility == ArgumentAttribute.VisibilityType.CmdLineOnly)
                     continue;
+                var name = inputAttr.Name ?? fieldInfo.Name;
+                // The order of GetFields is stable, meaning that
+                // fields are always returned in the same order and for this reason
+                // unit tests to compare manifest are passing. For the same reason
+                // duplicate name skipped are always in the same correct order.
+                // Same name field can bubble up from base class even though
+                // its overidden / hidden, skip it.
+                if (collectedFields.Contains(name))
```

Comment:
If there are two `Features` columns and the second one is hidden, which one will be picked up here? I couldn't see any mechanism to filter out hidden fields. Do I miss something?

Reply:
this are not about columns, this is about fields in a class

---
In reply to: [266176048](https://github.com/dotnet/machinelearning/pull/2964#discussion_r266176048) [](ancestors = 266176048)

Reply:
Sorry. Please replace `columns` with `fields` in my comment.

---
In reply to: [266176509](https://github.com/dotnet/machinelearning/pull/2964#discussion_r266176509) [](ancestors = 266176509,266176048)

Reply:
yes, i got it, pls see the comment i wrote on top

---
In reply to: [266176749](https://github.com/dotnet/machinelearning/pull/2964#discussion_r266176749) [](ancestors = 266176749,266176509,266176048)

Reply:
Do you think this comment is clearer?
```
                // The fields returned by GetFields is stably ordered starting from derived class' fields
                // and ending up with base class ones. Therefore,
                // unit tests to compare manifest are passing as long as the new field's type and name are identical to
                // what it overrides. For the same reason,
                // we can always skip the base filed overridden by another field introduced by "new".
```

---
In reply to: [266176749](https://github.com/dotnet/machinelearning/pull/2964#discussion_r266176749) [](ancestors = 266176749,266176509,266176048)

Reply:
Its not exactly correct. I am not sure that the order of fields are returned from derivatives and then for base classes. I didnt make such statement. Moreover in documentation for GetFields() its said that fields could be returned in any order. The only statements I am making: 1. Order is stable (meaning always the same order of fields) 2. For "currently duplicate" field (and its only one in our case NormalizeFields) order is "correct" with the fix, and that is the derivative is returned first and then the base. Note the difference, I do not state in general that derivative fields are returned first and then base one. 

---
In reply to: [266179947](https://github.com/dotnet/machinelearning/pull/2964#discussion_r266179947) [](ancestors = 266179947,266176749,266176509,266176048)

Reply:
talked to Wei-Sheng offline, this looks good to him as it is. Thank you

---
In reply to: [266546642](https://github.com/dotnet/machinelearning/pull/2964#discussion_r266546642) [](ancestors = 266546642,266179947,266176749,266176509,266176048)